### DISCUSSION TITLE
Vendor Spinel properties and Spinel interface access

### DISCUSSION THREAD

**Discussion by bertoldi-silabs** [openthread/openthread#9009] [Q&A] [answered]
Hello,

I am working on implementing a vendor specific set of Spinel handlers for properties specific to particular hardware, and I am running into two areas of concern on the posix side that I would greatly appreciate some insight and feedback on. For reference, this is based off of the ot-cli program from `src/posix/main.c` with an additional call to `otCliSetUserCommands` registering an externally linked list of command handlers.

### What is the intended way to access `ot::Posix::Radio::sRadioSpinel` in order to set and get properties from the RCP?

I found it very clear how to use the NCP vendor hook to set up the handlers on the secondary side, but it wasn't clear to me how to set up the complimentary structure on the posix side. The existing commands that use spinel appear to do so by having the cli interpreter forward the request to a platform API call, which is resolved in the posix radio class, which has `sRadioSpinel` in its scope. However, externally linked functions do not have scope access to the spinel interface, and I find that initializing a second spinel interface object usually results in contention with the hardware interface and its device descriptors.

The solution that I have had success with is adding an accessor function to ot::Posix::Radio which returns a void pointer to `sRadioSpinel`, as template type issues have raised errors with returning a more type appropriate reference. Would something like this seem appropriate for adding vendor spinel properties, or is there a differently intended way for the spinel interface to be accessed or properties to be registered?

### Should `ot::Spinel::RadioSpinel::GetWithParam` be supported as a normal use case beyond getting the RCP timestamp with dummy payload?

The structure of the spinel commands that I am implementing is based off command sets, (e.g. GPIO is a spinel property, write/read and direction are local property keys). This works without issue when using the `GetWithParam` function as opposed to the typical `Get`, but I noticed that the function has a comment stating that it should only be used in the previously stated case.

Would it be appropriate to use `GetWithParam` for the types of conditions I am working with, or should it only be reseved for getting the RCP timestamp?

Thank you very much for your consideration.

**Comment by abtink** [comment, marked as answer]
@bertoldi-silabs I can provide some suggestions which I hope may help.

> I found it very clear how to use the NCP vendor hook to set up the handlers on the secondary side, but it wasn't clear to me how to set up the complimentary structure on the posix side.

In general, the vendor spinel property support was intended for NCP model (that's why all example are for that). RCP was added later and vendor support for it was not there. So it is expected that the hooks for it are not there in `RadioSpinel` and posix host code.

> The solution that I have had success with is adding an accessor function to ot::Posix::Radio which returns a void pointer to sRadioSpinel, as template type issues have raised errors with returning a more type appropriate reference. Would something like this seem appropriate for adding vendor spinel properties, or is there a differently intended way for the spinel interface to be accessed or properties to be registered?

I think adding a `void *` can work (and casting it to right type on vendor code). It is not ideal model but I guess it may work.

----

I think your goal is to have vendor specific CLI commands that get/set vendor specific properties on RCP (radio). And you are considering to add new spinel vendor properties for this.

A different approach can be to use the existing diag mechanism.
- `RadioSpinel` on host posix side provides `otPlatDiagProcess()` which sends the command to RCP (using `SPINEL_PROP_NEST_STREAM_MFG` property). 
- On the RCP side, the `SPINEL_PROP_NEST_STREAM_MFG` prop handler will invoke `otDiagProcessCmdLine()` (now on RCP) to parse the command (which then passes it to `FactoryDiags::Diags` class).
-  The `FactoryDiags` will handle some basic command and if the command is not any of the basic ones,  it calls platform `otPlatDiagProcess()` to process it (this is now on RCP and its platform implementation). 
- We can potentially use this mechanism as way to interact with RCP from a posix host through the diag command and expose it as CLI commands to user.

**Comment by bertoldi-silabs** [comment]
Thanks for the thoughts @abtink.

I had not considered the distinction between the NCP and RCP models in this context, but it does make more sense.

You are correct in that right now I am working on vendor specific CLI commands which would need to get/set vendor specific properties on the RCP by interfacing with radio drivers. However, more specifically, I am working on adding an API to these drivers using spinel for applications to interface with, whether or not these are driven by a CLI.

Additionally I am considering the case in which users of our hardware wish to use this vendor API on an RCP system without including diag in the image. Even though the case can be made for Diag as being an open mechanism to communicate over spinel to the RCP, I think there can also be situations where these two functions would be considered as independent needs.

I also wouldn't consider the void pointer accessor an ideal model, but I do think that external access to RadioSpinel could be very useful. If there is a cleaner or more idiomatic way than a void pointer, I would be happy to consider it.
